API/Repositories/Implementations/PlayerRepository.cs

```diff
@@ -56,22 +56,44 @@ public async Task<IEnumerable<Player>> GetAsync(bool eagerLoad)
     }
 
     public async Task<IEnumerable<Player>> SearchAsync(string username) =>
-        await SearchQuery(username, true)
+        await _context.Players
+            .WhereUsername(username, true)
             .Include(p => p.Ratings)
             .AsNoTracking()
             .Take(30)
             .ToListAsync();
 
-    public async Task<Player?> GetAsync(string username)
+    public async Task<Player?> GetAsync(string username) =>
+        await _context.Players.WhereUsername(username, false).FirstOrDefaultAsync();
+
+    public async Task<Player?> GetVersatileAsync(string key, bool eagerLoad)
     {
-        List<Player> players = await SearchQuery(username, false).ToListAsync();
+        IQueryable<Player> query = _context.Players.AsNoTracking();
+        if (eagerLoad)
+        {
+            query = query.Include(p => p.User).ThenInclude(u => u!.Settings);
+        }
+
+        if (!int.TryParse(key, out var value))
+        {
+            return await query.WhereUsername(key, false).FirstOrDefaultAsync();
+        }
+
+        // Check for the player id
+        Player? result = await query.FirstOrDefaultAsync(p => p.Id == value);
+
+        if (result != null)
+        {
+            return result;
+        }
 
-        if (players.Count > 1)
+        // Check for the osu id
+        if (long.TryParse(key, out var longValue))
         {
-            throw new Exception("More than one player was found.");
+            return await query.FirstOrDefaultAsync(p => p.OsuId == longValue);
         }
 
-        return players.Count == 0 ? null : players.First();
+        return await query.WhereUsername(key, false).FirstOrDefaultAsync();
```

Comment:
I would delete the if statement on 77 since this is the default anyway.

Reply:
Yea I get the itch to do this every time I look at it but I was going to save it for a pull dedicated to code quality

Reply:
that's fine